leaderelection/leaderelection.go

```diff
@@ -0,0 +1,184 @@
+	"os"
+	"strings"
+	"sync"
+	"time"
+)
+
+var (
+	initOnce sync.Once
+	callOnce sync.Once
+	manager  *Manager
+)
+
+type ElectionManager interface {
+	IsLeader() bool
+	Run() error
+}
+
+type Manager struct {
+	log    *logp.Logger
+	ctx    context.Context
+	leader K8SLeaderElectionService
+	block  chan bool
+}
+
+// NewLeaderElector acts as a singleton
+func NewLeaderElector(ctx context.Context, log *logp.Logger, cfg config.Config) (ElectionManager, error) {
+	kubeClient, err := providers.KubernetesProvider{}.GetClient(cfg.KubeConfig, kubernetes.KubeClientOptions{})
+	if err != nil {
+		log.Errorf("NewLeaderElector error in GetClient: %v", err)
+		return nil, err
+	}
+
+	initOnce.Do(func() {
+		var leConfig le.LeaderElectionConfig
+		var leader K8SLeaderElectionService
+
+		block := make(chan bool)
+		leConfig, err = buildConfig(ctx, log, kubeClient, block)
+		leader, err = NewK8sLeaderElector(leConfig)
+
+		manager = &Manager{
+			log:    log,
+			ctx:    ctx,
+			leader: leader,
+			block:  block,
+		}
+	})
+
+	return manager, err
+}
+
+func GetLeaderElectorManager() ElectionManager {
+	return manager
+}
+
+func (m *Manager) IsLeader() bool {
+	return m.leader.IsLeader()
+}
+
+// Run leader election is blocking until a leader is being elected or timeout has reached.
+func (m *Manager) Run() error {
+	go m.leader.Run(m.ctx)
+	m.log.Infof("started leader election")
+
+	select {
+	case <-m.block:
+		m.log.Infof("new leader has been elected")
+	case <-time.After(FirstLeaderDeadline):
+		m.log.Warnf("timeout - no leader has been elected for %s", FirstLeaderDeadline.String())
+	}
+
+	return nil
+}
+
+func buildConfig(ctx context.Context, log *logp.Logger, client k8s.Interface, block chan bool) (le.LeaderElectionConfig, error) {
+	podId, err := currentPodID(log)
+	if err != nil {
+		return le.LeaderElectionConfig{}, err
+	}
+
+	id := fmt.Sprintf("%s_%s", LeaderLeaseName, podId)
+	ns, err := kubernetes.InClusterNamespace()
+	if err != nil {
+		ns = "default"
+	}
+
+	lease := metav1.ObjectMeta{
+		Name:      LeaderLeaseName,
+		Namespace: ns,
+	}
+
+	return le.LeaderElectionConfig{
+		Lock: &rl.LeaseLock{
+			LeaseMeta: lease,
+			Client:    client.CoordinationV1(),
+			LockConfig: rl.ResourceLockConfig{
+				Identity: id,
+			},
+		},
+		ReleaseOnCancel: true,
+		LeaseDuration:   LeaseDuration,
+		RenewDeadline:   RenewDeadline,
+		RetryPeriod:     RetryPeriod,
+		Callbacks: le.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				log.Infof("leader election lock GAINED, id: %v", id)
+			},
+			OnStoppedLeading: func() {
+				log.Infof("leader election lock LOST, id: %v", id)
+				// leaderelection.Run stops in case it has stopped holding the leader lease, we re-run the manager
+				// to keep following leader status.
+				go manager.leader.Run(ctx)
+				log.Infof("re-running leader elector")
+			},
+			OnNewLeader: func(identity string) {
+				log.Infof("leader election lock has been acquired, id: %v", identity)
```

Comment:
acquired by another cloudbeat

Reply:
right? because otherwise `OnStartedLeading` is called

Reply:
Both will be called, first the `OnNewLeader` and then the `OnStartedLeading` cb.

Reply:
add an `if identity == id` and write different and proper log messaged for these to cases (acquired by another, acquired by me)